internal/controller/postgrescluster/snapshots.go

```diff
@@ -133,49 +148,286 @@ func (r *Reconciler) reconcileVolumeSnapshots(ctx context.Context,
+	backupJobCompletionTime := backupJob.Status.CompletionTime.String()
+	// TODO: change this to check whether backupJobCompletionTime is more recent than pvc annotation.
+	if pvc.GetAnnotations()[naming.PGBackRestBackupJobCompletion] == backupJobCompletionTime {
+		return pvc, nil
+	}
+
+	// If we've made it here, the pvc has not been restored with latest backup.
+	// Find the dedicated snapshot volume restore job if it exists. Since we delete
+	// successful restores after we annotate the PVC and stop making restore jobs
+	// if a failed DSV restore job exists, there should only ever be one DSV restore
+	// job in existence at a time.
+	// TODO(dsessler7): Should this function throw an error or something if multiple
+	// DSV restores somehow exist?
+	restoreJob, err := r.getDedicatedSnapshotVolumeRestoreJob(ctx, cluster)
+	if err != nil {
+		return pvc, err
 	}
 
-	// Find pvc associated with primary instance
-	primaryPvc := corev1.PersistentVolumeClaim{}
-	for _, pvc := range clusterVolumes {
-		pvcInstance := pvc.GetLabels()[naming.LabelInstance]
-		pvcRole := pvc.GetLabels()[naming.LabelRole]
-		if pvcRole == naming.RolePostgresData && pvcInstance == primaryInstance.Name {
-			primaryPvc = pvc
+	// If we don't find a restore job, we run one.
+	if restoreJob == nil {
+		err = r.dedicatedSnapshotVolumeRestore(ctx, cluster, pvc, instances, clusterVolumes, backupJob)
+		return pvc, err
+	}
+
+	// If we've made it here, we have found a restore job. If the restore job was
+	// successful, set/update the annotation on the PVC and delete the restore job.
+	if restoreJob.Status.Succeeded == 1 {
+		if pvc.GetAnnotations() == nil {
+			pvc.Annotations = map[string]string{}
+		}
+		pvc.Annotations[naming.PGBackRestBackupJobCompletion] = restoreJob.GetAnnotations()[naming.PGBackRestBackupJobCompletion]
+		annotations := fmt.Sprintf(`{"metadata":{"annotations":{"%s": "%s"}}}`,
+			naming.PGBackRestBackupJobCompletion, pvc.Annotations[naming.PGBackRestBackupJobCompletion])
+
+		patch := client.RawPatch(client.Merge.Type(), []byte(annotations))
+		err = r.handlePersistentVolumeClaimError(cluster,
+			errors.WithStack(r.patch(ctx, pvc, patch)))
+
+		if err != nil {
+			return pvc, err
 		}
+
+		err = r.Client.Delete(ctx, restoreJob, client.PropagationPolicy(metav1.DeletePropagationBackground))
+		return pvc, errors.WithStack(err)
+	}
+
+	// If the restore job failed, create a warning event.
+	if restoreJob.Status.Failed == 1 {
+		r.Recorder.Event(cluster, corev1.EventTypeWarning,
+			"DedicatedSnapshotVolumeRestoreJobError", "restore job failed, check the logs")
+		return pvc, nil
+	}
+
+	// If we made it here, the restore job is still running and we should do nothing.
+	return pvc, err
+}
+
+// createDedicatedSnapshotVolume creates/updates/gets the dedicated snapshot volume.
+// It expects that the volume name and GVK has already been set on the pvc that is passed in.
+// TODO(dsessler7): The above fact makes this method feel a little awkward to me.
```

Comment:
If you feel it won't compromise the readability of `reconcileDedicatedSnapshotVolume`, go for it.